contrib/babelfishpg_tsql/src/procedures.c

```diff
@@ -2413,3 +2418,290 @@ sp_dropserver_internal(PG_FUNCTION_ARGS)
+		function_name_list = list_make2(makeString(physical_schema_name),makeString(function_name));
+		candidates = FuncnameGetCandidates(function_name_list, -1, NIL, false, false, false, true);
+
+		/* if no function is found */
+		if(candidates == NULL)
+			ereport(ERROR,
+					(errcode(ERRCODE_UNDEFINED_OBJECT),
+					errmsg("function does not exist")));
+
+		/* check if the current user has priviledge on the function */
+		if(pg_proc_aclcheck(candidates->oid, user_id, ACL_EXECUTE) != ACLCHECK_OK)
+			ereport(ERROR,
+					(errcode(ERRCODE_SYNTAX_ERROR),
+					errmsg("current user does not have priviledges on the function")));
+		
+		/* check if multiple function with same function name exits */
+		if(candidates->next != NULL)
+			ereport(ERROR,
+					(errcode(ERRCODE_SYNTAX_ERROR),
+					errmsg("multiple functions with same function name exits")));
+
+		function_id = candidates->oid;
+		full_function_name = psprintf("\"%s\".\"%s\"", physical_schema_name, function_name);
+		function_signature = (char *)get_pltsql_function_signature_internal(full_function_name, candidates->nargs, candidates->args);
+
+		list_free(function_name_list);
+		pfree(candidates);
+		pfree(full_function_name);
+		pfree(physical_schema_name);
+	}
+	
+	/*
+	 * If both volatility and function name is not provided then it will return a list of functions present in current database.
+	 * else if only volatility is not provided the it will return the volatility of the specified function.
+	 * If both volatility and function name is provided it will set the function volatility to the specified volatility
+	 */
+	if(volatility == NULL)
+	{	
+		if(function_name == NULL)
+		{	
+			query = psprintf(
+					"SELECT t3.orig_name as SchemaName, t1.proname as FunctionName, "
+					"CASE "
+						"WHEN t1.provolatile = 'v' THEN 'volatile' "
+						"WHEN t1.provolatile = 's' THEN 'stable' "
+						"ELSE 'immutable' "
+					"END AS Volatility "
+					"from (SELECT (aclexplode(proacl)).*, proname, provolatile, pronamespace from pg_proc WHERE prokind = 'f') t1 "
+					"JOIN pg_namespace t2 ON t1.pronamespace = t2.oid "
+					"JOIN sys.babelfish_namespace_ext t3 ON t3.nspname = t2.nspname "
+					"where t1.grantee = %d AND t3.dbid = sys.db_id() ORDER BY t3.orig_name, t1.proname", user_id
+				);
+		}
+		else
+		{
+			query = psprintf(
+					"SELECT t3.orig_name as SchemaName, CAST('%s' as sys.varchar) as FunctionName, "
+					"CASE "
+						"WHEN provolatile = 'v' THEN 'volatile' "
+						"WHEN provolatile = 's' THEN 'stable' "
+						"ELSE 'immutable' "
+					"END AS Volatility from pg_proc t1 "
+					"JOIN pg_namespace t2 ON t1.pronamespace = t2.oid "
+					"JOIN sys.babelfish_namespace_ext t3 ON t3.nspname = t2.nspname "
+					"where t1.oid = %u", function_name, function_id
+				);
+		}
+
+		PG_TRY();
+		{	
+			char nulls = 0;
+			MemoryContext savedPortalCxt;
+			SPIPlanPtr plan;
+			Portal portal;
+			DestReceiver *receiver;
+
+			savedPortalCxt = PortalContext;
```

Comment:
this could be problematic in cases when there is error, who will reset portal context?

Reply:
I have reset the portal context on line 2632.

Reply:
but you have error condition at 2629

Reply:
Have corrected it in the latest commit